py-polars/polars/utils/udfs.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import dis
+import sys
+import warnings
+from inspect import signature
+from itertools import tee
+from typing import Any, Callable, Iterable
+
+from polars.exceptions import PolarsInefficientApplyWarning
+from polars.utils.various import find_stacklevel
+
+# Note: in 3.11 individual binary opcodes were folded into a new BINARY_OP
+_BINARY_OPCODES = {
+    "BINARY_ADD": "+",
+    "BINARY_AND": "&",
+    "BINARY_FLOOR_DIVIDE": "//",
+    "BINARY_MODULO": "%",
+    "BINARY_MULTIPLY": "*",
+    "BINARY_OR": "|",
+    "BINARY_POWER": "**",
+    "BINARY_SUBTRACT": "-",
+    "BINARY_TRUE_DIVIDE": "/",
+    "BINARY_XOR": "^",
+}
+_SIMPLE_EXPR_OPS = {
+    "BINARY_OP",
+    "COMPARE_OP",
+    "CONTAINS_OP",
+    "IS_OP",
+    "LOAD_CONST",
+    "LOAD_FAST",
+    "UNARY_NOT",
+}
+_SIMPLE_FRAME_OPS = _SIMPLE_EXPR_OPS | {"BINARY_SUBSCR"}
+
+
+def _pairwise(values: Iterable[Any]) -> Iterable[tuple[Any, Any]]:
+    a, b = tee(values)
+    next(b, None)
+    return zip(a, b)
+
+
+def _expr(value: str | tuple[str, str, str], col: str | None) -> str:
+    if isinstance(value, tuple):
+        op = value[1]
+        if len(value) == 2 and op == "not":
+            return f"~{_expr(value[0], col)}"
+        elif value[2] == "None" and op.startswith("is"):
+            null_check = "is_not_null" if value[1].endswith("not") else "is_null"
+            return f"{_expr(value[0], col)}.{null_check}()"
+        elif op.endswith("in"):
+            not_ = "~" if op.startswith("not") else ""
+            return f"({not_}{_expr(value[0], col)}.is_in({_expr(value[2], col)}))"
+        else:
+            return f"({_expr(value[0], col)} {op} {_expr(value[2], col)})"
+
+    elif col and value.isidentifier():
+        if value not in ("True", "False", "None"):
+            return f'pl.col("{col}")'
+
+    return value
+
+
+def _op(op: str, argrepr: str, argval: Any) -> str:
+    if argrepr:
+        return argrepr
+    elif op == "IS_OP":
+        return "is not" if argval else "is"
+    elif op == "CONTAINS_OP":
+        return "not in" if argval else "in"
+    elif op == "UNARY_NOT":
+        return "not"
+    raise AssertionError(
+        "Unreachable code - please report a bug to https://github.com/pola-rs/polars/issues "
+        "with the content of function you were passing to `apply`."
+    )
+
+
+def _get_bytecode_ops(function: Callable[[Any], Any]) -> list[tuple[str, str, Any]]:
+    """Return disassembled bytecode ops, arg-repr, and arg-specific value/flag."""
+    try:
+        instructions = list(dis.get_instructions(function))
+        idx_last = len(instructions) - 1
+        return [
+            _inst(inst.opname, inst.argrepr, inst.argval)
+            for idx, inst in enumerate(instructions)
+            if (idx, inst.opname) not in ((0, "RESUME"), (idx_last, "RETURN_VALUE"))
+        ]
+    except TypeError:
+        return []  # can't disassemble non-native functions (eg: numpy/etc)
```

Comment:
there are a bunch of numpy functions which people might use though (e.g. `np.sin`), not knowing that there's an equivalent

perhaps this is a case where we can consider (later) falling back to `inspect.getsource`, so as to be able to make a suggestion there too?

Reply:
Actually... I think I put this note there for a reason other than numpy - it's a bit misleading 😅 We probably could actually identify numpy functions by looking for the following combination:

`LOAD_GLOBAL` opcode  associated with "np" or "numpy" argvalue.
`LOAD_METHOD` opcode with an argvalue representing a function we support (eg: "sin").